File: web-service/security/src/main/java/io/ehdev/conrad/security/filter/StatelessAuthenticationFilter.java
package io.ehdev.conrad.security.filter;

import io.ehdev.conrad.security.database.model.UserModel;
import io.ehdev.conrad.security.database.repositories.UserModelRepository;
import io.ehdev.conrad.security.jwt.JwtManager;
import io.ehdev.conrad.security.jwt.UserToken;
import io.ehdev.conrad.security.user.UserCookieManger;
import io.ehdev.conrad.security.user.UserPrincipal;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Collection;
import java.util.Collections;
import java.util.UUID;

@Component
public class StatelessAuthenticationFilter extends OncePerRequestFilter {

    private static final Logger logger = LoggerFactory.getLogger(StatelessAuthenticationFilter.class);

    static final String HEADER_NAME = "X-AUTH-TOKEN";

    private final UserModelRepository userModelRepository;
    private final UserCookieManger userCookieManger;
    private final JwtManager jwtManager;

    @Autowired
    public StatelessAuthenticationFilter(UserModelRepository userModelRepository,
                                         UserCookieManger userCookieManger,
                                         JwtManager jwtManager) {
        this.userModelRepository = userModelRepository;
        this.userCookieManger = userCookieManger;
        this.jwtManager = jwtManager;
    }

    String findTokenString(HttpServletRequest request) {
        if(request.getHeader(HEADER_NAME) != null) {
            return request.getHeader(HEADER_NAME);
        } else {
            return userCookieManger.readCookieValue(request);
        }
    }

    Authentication processUserToken(UserToken userToken) {
        if(userToken == null) {
            return null;
        }

        UserModel user = userModelRepository.findOne(UUID.fromString(userToken.getUserId()));

        if(user == null) {
            return null;
        }

        logger.debug("Logging in {}", user.getId());
        UserPrincipal userPrincipal = new UserPrincipal(user);
        Collection<GrantedAuthority> authorities = Collections.singleton(new SimpleGrantedAuthority("ROLE_USER"));
        return new UsernamePasswordAuthenticationToken(userPrincipal, null, authorities);
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {

        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        logger.debug("Authentication: {}", authentication);
        if (authentication == null || !authentication.isAuthenticated()) {
            String tokenValue = findTokenString(request);

            logger.debug("tokenValue: {}", tokenValue);

            if (StringUtils.isNotBlank(tokenValue)) {
                UserToken userToken = jwtManager.parseToken(tokenValue);
                logger.debug("User token from cookie: {}", userToken);
                SecurityContextHolder.getContext().setAuthentication(processUserToken(userToken));
            }
        }
        filterChain.doFilter(request, response);
    }
}

File: auth/authentication/src/main/kotlin/tech/crom/security/authentication/config/stateless/StatelessAuthenticationFilter.kt
package tech.crom.security.authentication.config.stateless

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service
import org.springframework.web.filter.GenericFilterBean
import tech.crom.database.api.RepoManager
import tech.crom.database.api.UserManager
import tech.crom.model.security.authentication.CromRepositoryAuthentication
import tech.crom.model.security.authentication.CromUserAuthentication
import tech.crom.security.authentication.cookie.UserCookieManger
import tech.crom.security.authentication.jwt.JwtManager
import tech.crom.security.authentication.jwt.JwtTokenAuthentication
import javax.servlet.FilterChain
import javax.servlet.ServletRequest
import javax.servlet.ServletResponse
import javax.servlet.http.HttpServletRequest

@Service
class StatelessAuthenticationFilter @Autowired constructor(
    val cookieManager: UserCookieManger,
    val jwtManager: JwtManager,
    val userManager: UserManager,
    val repoManager: RepoManager): GenericFilterBean() {

    override fun doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain) {
        setAuthenticationFromHeader(request as HttpServletRequest)
        chain.doFilter(request, response)
    }

    private fun setAuthenticationFromHeader(request: HttpServletRequest) {
        val authentication: Authentication? = SecurityContextHolder.getContext().authentication

        if(authentication != null) {
            val cookieValue = cookieManager.readCookieValue(request) ?: return
            val token = jwtManager.parseToken(cookieValue) ?: return

            if (token is JwtTokenAuthentication.UserJwtTokenAuthentication) {
                val user = userManager.findUserDetails(token.userUuid) ?: return
                SecurityContextHolder.getContext().authentication = CromUserAuthentication(user)
            } else if (token is JwtTokenAuthentication.RepoJwtTokenAuthentication) {
                val repo = repoManager.findRepo(token.repoUuid) ?: return
                SecurityContextHolder.getContext().authentication = CromRepositoryAuthentication(repo)
            }
        }
    }
}
